src/os/bluestore/BlueStore.cc

```diff
@@ -1695,6 +1695,12 @@ int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint64_t offset, u
     if (b->offset >= end) {
       break;
     }
+    // If no overlap is found between what we want to read and what we have
+    // cached, then forget about continuing
+    if (std::max(b->end(), offset + length) - std::min(b->offset, offset) >
+        length + b->length) {
+      break;
+    }
```

Comment:
`i` from `_data_lower_bound()` either gives buffer that contains offset, or first one after.
I do not understand what this condition is doing.
Did you have in mind a case that `b->offset < end` but also `b->offset + b->length < offset` ?
That would be assert against `_data_lower_bound()` guarantee.

Reply:
This disjointment test seems unnecessary.

Reply:
I've modified it to work with overlapping functions (the previous formula was terrible idk how I got past kindergarten math courses), imho it's easier to think about overlapping ranges.